python/ray/tune/trainable.py

```diff
@@ -440,12 +440,17 @@ def _export_model(self, export_formats, export_dir):
         """
         return {}
 
-
+# TODO(gehring): refactor `wrap_function` and move it to function_runner.py
+#     since the what this function should do is completely dependent on the
+#     how `FunctionRunner` is implemented.
 def wrap_function(train_func):
     from ray.tune.function_runner import FunctionRunner
 
     class WrappedFunc(FunctionRunner):
-        def _trainable_func(self):
-            return train_func
+
+        def _trainable_func(self, config, reporter):
+            output = train_func(config, reporter)
+            reporter(done=True)
+            return output
```

Comment:
In order for this to be compatible with existing trainable functions, I believe `_trainable_func` would need to return a function rather than the output of the called function. Are you proposing an alternative to wrapping functions as Trainable objects for logging?

Reply:
The `_trainable_func` method belongs to the `FunctionWrapper` class. As I understand it, it is not part of the `Trainable` class API. In both the old and new implementation, this method is used to dynamically create a `Trainable` class which uses the provided function.

This change is only to simplify how functions are wrapped within a Trainable object. I felt it was clunky and unnecessary to have to call this method "twice", e.g., `_trainable_func()(config, reporter)`. Alternatively, if we would want to keep it such that it returns a function, we could consider making it a property (but then again, this feels unnecessary).

Additionally, I am getting ready to propose a generator based function API which would simplify things greatly. With this new way of wrapping the function, backwards compatibility should be clean and straightforward.

This new implementation of ray's FunctionRunner accounts for these change so shouldn't cause any problems. This should not affect any older versions unless someone tries to unpickle incompatible versions or tries to access the "private" method directly. Even then, I would think it unlikely as the FunctionWrapper instances aren't exposed through the public API and users should only be using one version of ray at a time.

That being said, this change is minor and completely optional. We can revert back to the old way but I am not convinced this is necessary. Given the re-write of `FunctionWrapper`, this would be a good time to change it.